akka-http2-support/src/main/scala/akka/http/impl/engine/http2/RequestParsing.scala

```diff
@@ -19,69 +23,84 @@ import scala.collection.immutable.VectorBuilder
-
-        val entity =
-          if (subStream.data == Source.empty || contentLength == 0) HttpEntity.Empty
-          else if (contentLength > 0) HttpEntity.Default(contentType, contentLength, subStream.data)
-          else HttpEntity.Chunked.fromData(contentType, subStream.data)
-
-        val uri = Uri(scheme, authority).withPath(Uri.Path(path))
-        HttpRequest(
-          method, uri, headers.result(), entity, HttpProtocols.`HTTP/2.0`
-        )
-      } else remainingHeaders.head match {
-        case (":scheme", value) ⇒
-          rec(remainingHeaders.tail, method, value, authority, path, contentType, contentLength, headers)
-        case (":method", value) ⇒
-          val m = HttpMethods.getForKey(value).getOrElse(malformedRequest(s"Unknown HTTP method: '$value'"))
-          rec(remainingHeaders.tail, m, scheme, authority, path, contentType, contentLength, headers)
-        case (":path", path) ⇒
-          rec(remainingHeaders.tail, method, scheme, authority, path, contentType, contentLength, headers)
-        case (":authority", value) ⇒
-          val authority = Uri.Authority.parse(value)
-          rec(remainingHeaders.tail, method, scheme, authority, path, contentType, contentLength, headers)
-
-        case ("content-type", ct) ⇒
-          val contentType = ContentType.parse(ct).right.getOrElse(malformedRequest(s"Invalid content-type: '$ct'"))
-          rec(remainingHeaders.tail, method, scheme, authority, path, contentType, contentLength, headers)
-
-        case ("content-length", length) ⇒
-          val contentLength = length.toLong
-          rec(remainingHeaders.tail, method, scheme, authority, path, contentType, contentLength, headers)
-
-        case (name, value) ⇒
-          // FIXME: later modify by adding HttpHeaderParser.parseHttp2Header that would use (name, value) pair directly
-          //        or use a separate, simpler, parser for Http2
-          // FIXME: add correctness checks (e.g. duplicated content-length) modeled after ones in HttpMessageParser
-
-          // The odd-looking 'x' below is a by-product of how current parser and HTTP/1.1 work.
-          // Without '\r\n\x' (x being any additional byte) parsing will fail. See HttpHeaderParserSpec for examples.
-          val concHeaderLine = name + ": " + value + "\r\nx"
-
-          httpHeaderParser.parseHeaderLine(ByteString(concHeaderLine))()
-          rec(remainingHeaders.tail, method, scheme, authority, path, contentType, contentLength, headers += httpHeaderParser.resultHeader)
-      }
-
-    rec(subStream.initialHeaders.keyValuePairs)
+
+  def parseRequest(httpHeaderParser: HttpHeaderParser, serverSettings: ServerSettings, attributes: Attributes): Http2SubStream ⇒ HttpRequest = {
+    val remoteAddressHeader =
+      if (serverSettings.remoteAddressHeader) {
+        // in order to avoid searching all the time for the attribute, we need to guard it with the setting condition
+        attributes.get[HttpAttributes.RemoteAddress] match {
```

Comment:
While you are at it, you can fix `HttpAttributes.RemoteAddress.address` to be an `InetSocketAddress` instead of an `Option[InetSocketAddress]`. The option does not seem to serve any purpose. This code can then be just

```scala
attributes.get[HttpAttributes.RemoteAddress].map(remote => model.headers.`Remote-Address`(RemoteAddress(remote.address)))
```

(Actually, no need for `OptionVal` here, sorry, did only notice it while reading it another time)